File: 23602.py
from math import hypot as h

A, B, C = map(int,input().split())
a, b, c = map(int,input().split())

K = C/(A*A + B*B)
k = c/(a*a + b*b)

P, Q = K*A, K*B
p, q = k*a, k*b

H = lambda a, b, c, d: h(a, b) + h(c, d) + h(a - c, b - d)

if a*B == A*b:
    print(H(P, Q, p, q))
    exit()

x = (B*c - b*C)/(B*a - b*A)
y = (a*C - A*c)/(B*a - b*A)

def d(I, i):
    
    S = (I*x + (1 - I)*P)
    T = (I*y + (1 - I)*Q)
    s = (i*x + (1 - i)*p)
    t = (i*y + (1 - i)*q)
    
    return H(S, T, s, t)

u = [[0,0],[1,0],[1,1],[0,1],[-1,1],[-1,0],[-1,-1],[0,-1],[1,-1]]
I = i = 0.5
for z in range(20):
    w = 2**(-2 - z)
    while 1:
        l = [d(I + v[0]*w, i + v[1]*w) for v in u]
        m = min(l)
        if m != l[0]:
            X, Y = u[l.index(m)]
            I += X*w
            i += Y*w
        else:
            break

print(d(I, i))

File: 23602.ml
Scanf.scanf "%f %f %f\n%f %f %f\n" @@ fun aa bb cc a b c ->
  let (kk, k) = (cc/.(aa*.aa +. bb*.bb), c/.(a*.a +. b*.b)) in
  let (pp, qq, p, q) = (kk*.aa, kk*.bb, k*.a, k*.b) in
  let h a b c d = hypot a b +. hypot c d +. hypot (a -. c) (b -. d) in
  if a*.bb = aa*.b then print_float @@ h pp qq p q else
    let (x, y) = ((bb*.c -. b*.cc)/.(bb*.a -. b*.aa), (a*.cc -. aa*.c)/.(bb*.a -. b*.aa)) in
    let d ii i =
      let (ss, tt, s, t) = (
        ii*.x +. (1. -. ii)*.pp,
        ii*.y +. (1. -. ii)*.qq,
        i*.x +. (1. -. i)*.p,
        i*.y +. (1. -. i)*.q
      ) in h ss tt s t in
    let u = [
      (0., 0.); (1., 0.); (1., 1.); (0., 1.);
      (-1., 1.); (-1., 0.); (-1., -1.); (0., -1.); (1., -1.)] in
    let rec pow = function 0 -> 0.25 | z -> (pow @@ z - 1)/.2. in
    let rec grad_desc ii i z =
      let w = pow z in
      let rec listify = function
      | [] -> []
      | h :: t -> d (ii +. w*.(fst h)) (i +. w*.(snd h)) :: listify t in
      let l = listify u in
      let rec minl = function
      | [] -> infinity
      | h :: t -> min h @@ minl t in
      let m = minl l in
      let rec find x = function
        | [] -> -1
        | h :: t -> if x = h then 0 else 1 + find x t in
      let index = find m l in
      let (xx, yy) = let pair = List.nth u index in (fst pair, snd pair) in
      if m = List.nth l 0 then
        if z = 20 then d ii i else grad_desc ii i @@ z + 1
      else grad_desc (ii +. xx*.w) (i +. yy*.w) z in
      grad_desc 0.5 0.5 0 |> print_float
